Feature/kube improvements

olegTarassov opened the pull request:
1) Test KUBECONFIG global variable instead of file
- location of kube/config can be moved or appended using several files
2) Avoid printing kubernetes cluster version unless needed
- keep terminal line to the minimal
3) Put some nice colors
- control context or namespace color
4) Add Separator for geometry_kube

Commit:
Added default Variables
- Control display of k8s version
- Control color of namespace / context

Commit:
kube enhancement
1) Test KUBECONFIG global variable instead of file
- location of kube/config can be moved or appended using several files
2) Avoid printing kubernetes cluster version unless needed
- keep terminal line to the minimal
3) Put some nice colors
- control context or namespace color
4) Add Separator for geometry_kube

Commit:
type + added separator to the list.

Commit:
Merge branch 'mnml' into feature/kube_improvements

jedahan commented on functions/geometry_kube.zsh:
```diff
@@ -3,15 +3,28 @@
-  local context="$(kubectl config current-context 2> /dev/null)"
-  local namespace="$(kubectl config view -o "jsonpath={.contexts[?(@.name==\"$context\")].context.namespace}" 2> /dev/null)"
+  local geometry_context="$(kubectl config current-context 2> /dev/null)"
+  local geometry_namespace="$(kubectl config view -o "jsonpath={.contexts[?(@.name==\"$context\")].context.namespace}" 2> /dev/null)"
```
maybe `kube_context` and `kube_namespace`?

jedahan replied:
Though I can see an argument for this being geometry specific, not kube_specific. Either way, these variables should be 'hidden' unless something breaks, in which case letting the person know 'this is not kube broken, its your prompt' is less panic-inducing.

jedahan commented on functions/geometry_kube.zsh:
```diff
@@ -3,15 +3,28 @@
+    geometry_context="$(ansi ${GEOMETRY_KUBE_CONTEXT_COLOR:=white} ${geometry_context})"
+  fi
+
+  if [[ ! -z "${GEOMETRY_KUBE_NAMESPACE_COLOR}" ]]; then
```
does this make `${GEOMETRY_KUBE_NAMESPACE_COLOR:=white}` redundant?

I think this entire test and else branch can be removed and have the same effect?

olegTarassov replied:
You are correct, the default white is redundant.

Secondly, technically we can remove the if else and just have:

```bash
geometry_namespace="$(ansi ${GEOMETRY_KUBE_NAMESPACE_COLOR:=white} ${geometry_namespace:=default})"
```

My reason for the "if / else" is because your terminal  foreground colour might not be white like an off-white or grey and the default white would overwrite this unwillingly.

This ways if you dont add `GEOMETRY_KUBE_NAMESPACE_COLOR` or `GEOMETRY_KUBE_CONTEXT_COLOR`, you can keep terminal profile colour settings. 

I guess this would be the proposed:

```bash
  if [[ ! -z "${GEOMETRY_KUBE_CONTEXT_COLOR}" ]]; then
    geometry_context="$(ansi ${GEOMETRY_KUBE_CONTEXT_COLOR} ${geometry_context})"
  fi

  if [[ ! -z "${GEOMETRY_KUBE_NAMESPACE_COLOR}" ]]; then
    geometry_namespace="$(ansi ${GEOMETRY_KUBE_NAMESPACE_COLOR} ${geometry_namespace:=default})"
  fi

# [...]

echo -n "${geometry_kube_separator}${geometry_kube} ${geometry_context}:${geometry_namespace:=default}${geometry_kube_separator}"
```

jedahan replied:
I'd also be open to switching `ansi`'s argument order, and making the second argument optional. This would let us possibly remove all of the 'white' defaults easier.

jedahan replied:
though that kinda messes with the ability to have text with spaces. Which ansi breaks right now too I guess?

jedahan replied:
Okay, white can be changed to `default`

olegTarassov replied:
yes ansi will break if empty.
the geometry_namespace needs the default because kubernetes namespace returns empty string sometimes when namespace is default. This behaviour is not seen observed on the context.

jedahan commented on options.md:
```diff
@@ -11,19 +11,19 @@ and functionality.
   - [`geometry_hostname`](#geometryhostname)
   - [`geometry_jobs`](#geometryjobs)
-  - [`geometry_kube`](#geometrykube)
+  - [`geometry_kube`](#geometry_kube)
```
nice catch!

jedahan commented on functions/geometry_kube.zsh:
```diff
@@ -3,15 +3,28 @@
+    geometry_namespace="${geometry_namespace:=default}"
+  fi
+
+  echo -n "${geometry_kube_separator}${geometry_kube} ${geometry_context}:${geometry_namespace}${geometry_kube_separator}"
```
So whats neat is that zsh supports joins that are a bit smarter - you can look at `geometry::wrap` for how it works, but if you have an array of things you want to join with a string, you can do something like (untested)
```zsh
geometry_kube_info=($geometry_kube ${geometry_context}:${geometry_namespace})
echo "${(psj.${geometry_kube_separator}.)geometry_kube_info}"
```

olegTarassov replied:
let me try this out

Commit:
Added Improvements and PR Recommendations

- Group Variable declaration
- Append kube info to array
- Added ZSH parameter expansion

jedahan commented:
Thanks for the updates, I rebased and merged in #266
